tools/tertiary-analysis/scanpy/anndata_operations.xml

```diff
@@ -61,6 +77,47 @@ else:
     logging.warning('No genes starting with {} found, skip calculating expression of {} genes'.format('${flag.startswith}', '${flag.flag}'))
 #end for
 
+#if $copy_r.default and $copy_r.r_source:
+ad_s = sc.read('r_source.h5')
+if not all(adata.obs.index.isin(ad_s.obs.index)):
+  logging.error("Specified object for .raw must contain all .obs from main object.")
+  sys.exit()
+else:
+  adata.raw = ad_s[adata.obs.index]
+#end if
+
+#if $copy_x.default and len($copy_x.xlayers) > 0:
+#for $i, $x_s in enumerate($copy_x.xlayers):
+ad_s = sc.read('x_source_${i}.h5')
+if adata.n_obs == ad_s.n_obs and all(adata.obs_names == ad_s.obs_names):
+  #set xs=$copy_x.xlayers[$i]
+  if "${xs.dest}" == '': 
+    logging.error("%sth destination layer for %sth X source not specified" % ("${i}", "${i}"))
+    sys.exit()
+  adata.layers["${xs.dest}"] = ad_s.X
+else:
+  logging.warning("X source ${i} AnnData file is not compatible to be merged to main AnnData file, different cell names.")
```

Comment:
I wonder if this should be an error, and throw and exit code, rather than a warning, otherwise the user will think in the interface that this part went through. Hard to tell given that we are iterating over various objects....

Reply:
I was wondering why you didn't write these as fatal errors at other steps...